src/adapters/gumgum.js

```diff
@@ -0,0 +1,104 @@
+const bidmanager = require('../bidmanager');
+const utils      = require('../utils');
+const adloader   = require('../adloader');
+
+const BIDDER_CODE = 'gumgum';
+const CALLBACKS = {};
+
+const GumgumAdapter = function GumgumAdapter() {
+
+  const bidEndpoint = `https://g2.gumgum.com/hbid/imp`;
+
+  const WINDOW = global.top;
+  const SCREEN = WINDOW.screen;
+
+  function _callBids({ bids }) {
+    const browserParams = {
+      vw: WINDOW.innerWidth,
+      vh: WINDOW.innerHeight,
+      sw: SCREEN.width,
+      sh: SCREEN.height,
+      pu: WINDOW.location.href,
+      dpr: WINDOW.devicePixelRatio || 1
+    };
+    utils._each(bids, bidRequest => {
+      const { bidId
+            , params = {}
+            , placementCode
+            } = bidRequest;
+      const trackingId = params.inScreen;
+      const nativeId   = params.native;
+      const slotId     = params.inSlot;
+      const bid = {};
+
+      /* slot/native ads need the placement id */
+      switch (true) {
+        case !!(params.inImage):  bid.pi = 1; break;
+        case !!(params.inScreen): bid.pi = 2; break;
+        case !!(params.inSlot):   bid.pi = 3; break;
+        case !!(params.native):   bid.pi = 5; break;
+        default: return utils.logWarn(
+          `[GumGum] No product selected for the placement ${placementCode}` +
+          ', please check your implementation.'
+        );
+      }
+      /* tracking id is required for in-image and in-screen */
+      if (trackingId) bid.t = trackingId;
+      /* native ads require a native placement id */
+      if (nativeId) bid.ni = nativeId;
+      /* slot ads require a slot id */
+      if (slotId) bid.si = slotId;
+
+      const cachedBid = Object.assign({
+        placementCode,
+        id: bidId
+      }, bid);
+
+      const callback = { jsonp: `$$PREBID_GLOBAL$$.handleGumGumCB['${ bidId }']` };
+      CALLBACKS[bidId] = _handleGumGumResponse(cachedBid);
+      const query = Object.assign(callback, browserParams, bid);
+      const bidCall = `${bidEndpoint}?${utils.parseQueryStringParameters(query)}`;
+      adloader.loadScript(bidCall);
+    });
+  }
+
+  const _handleGumGumResponse = cachedBidRequest => bidResponse => {
+    const ad = bidResponse && bidResponse.ad;
+    if (ad && ad.id) {
+      const bid = bidfactory.createBid(1);
+      const { t: trackingId
+            , pi: productId
+            , placementCode
+            } = cachedBidRequest;
+      bidResponse.placementCode = placementCode;
+      const encodedResponse = encodeURIComponent(JSON.stringify(bidResponse));
+      const gumgumAdLoader = `r("${trackingId}",${productId},"${ encodedResponse }",window,top)`;
+      const gumgumLibrary = `(function(r){${gumgumAdLoader}})(function(trackingId,prodId,data,context,w,d,s,G){` +
+        'G=w.GUMGUM;d=w.document;function lg(){w.GUMGUM.pbjs(trackingId,prodId,data,context,this)}return!G?w' +
+        '.$$PREBID_GLOBAL$$.loadScript("https://g2.gumgum.com/javascripts/ggv2.js",lg):lg()});';
```

Comment:
A lot of our creatives assume the GumGum script is running on `top`, so this function is intended to pass our winning bid, to the entry point in our script (`top.GUMGUGM.pbjs`), but first we need to be sure our script exists on `top`.

Here's the uncompressed script:

```javascript
(function (request) {
    request( /* tracking id */ , /* productId */ , /* ad payload */ , /* iframed window */ , /* top window*/ );
})(function (trackingId, prodId, data, context, w, d, s, G) {
    G = w.GUMGUM;
    d = w.document;
    // load gumgum script
    function lg() {
        w.GUMGUM.pbjs(trackingId, prodId, data, context, this);
    }
    return !G
        // if the gumgum object doesn't exist on the top window, we need to load it
        // and then pass the payload
        ? w.$$PREBID_GLOBAL$$.loadScript("https://g2.gumgum.com/javascripts/ggv2.js", lg)
        // if the script is already running, just pass the payload
        : lg();
});
```